README.md
# Docker Image for Open Swoole

> This repo works with Open Swoole since release version v4.7.1. WIP

[![Tests](https://github.com/openswoole/docker-swoole/workflows/Tests/badge.svg)](https://github.com/openswoole/docker-swoole/actions)
[![Docker Pulls](https://img.shields.io/docker/pulls/openswoole/swoole.svg)](https://hub.docker.com/r/openswoole/swoole)
[![License](https://img.shields.io/badge/license-apache2-blue.svg)](https://github.com/openswoole/docker-swoole/blob/master/LICENSE)

This image is built for general-purpose. We have different examples included in this Git repository to help developers
to get familiar with the image and _Swoole_.

You can get the image from [here](https://hub.docker.com/r/openswoole/swoole).

Table of Contents
=================

   * [Feature List](#feature-list)
   * [How to Use This Image](#how-to-use-this-image)
      * [How to Install More PHP Extensions](#how-to-install-more-php-extensions)
      * [More Examples](#more-examples)
   * [Image Variants](#image-variants)
   * [Supported Tags and Respective Dockerfile Links](#supported-tags-and-respective-dockerfile-links)
      * [Versioned images](#versioned-images-based-on-stable-releases-of-swoole)
         * [Swoole 4.7](#swoole-47)
      * [Nightly images](#nightly-images-built-daily-using-the-master-branch-of-swoole-src)
   * [Build Images Manually](#build-images-manually)
   * [Credits](#credits)

# Feature List

* Built-in scripts to manage _Swoole_ extensions and _Supervisord_ programs.
* Easy to manage booting scripts in Docker.
* Allow running PHP scripts and other commands directly in different environments (including ECS).
* Use one root filesystem for simplicity (one Docker `COPY` command only in dockerfiles).
* _Composer_ included (_Composer v1_ for Swoole 4.5.8 and before; _Composer v2_ for Swoole 4.5.9 and after).
* Built for different architectures.
* Support auto-reloading for local development.
* Support code debugging for local development.

# How to Use This Image

The `openswoole/swoole` image is built using [the official PHP image](https://hub.docker.com/_/php) as base image, with a few changes.
For basic usage, please check the description section of [the official PHP image](https://hub.docker.com/_/php).

## How to Install More PHP Extensions

Same as in the official PHP image, most PHP extensions can be installed/configured using built-in helper scripts `docker-php-ext-configure`, `docker-php-ext-install`, `docker-php-ext-enable`, and `docker-php-source`. Here are some examples.

```Dockerfile
# To install the MySQL extensions.
FROM openswoole/swoole:4.7-php7.4-alpine

RUN docker-php-ext-install mysqli pdo_mysql
```

```Dockerfile
# To install the Redis extension.
FROM openswoole/swoole:4.7-php7.4-alpine

RUN set -ex \
    && pecl update-channels \
    && pecl install redis-stable \
    && docker-php-ext-enable redis
```

```Dockerfile
# To install the Couchbase extension.
FROM openswoole/swoole:4.7-php7.4-alpine

RUN set -ex \
    && apk update \
    && apk add --no-cache libcouchbase=2.10.6-r0 \
    && apk add --no-cache --virtual .build-deps $PHPIZE_DEPS libcouchbase-dev=2.10.6-r0 zlib-dev \
    && pecl update-channels \
    && pecl install couchbase-2.6.2 \
    && docker-php-ext-enable couchbase \
    && apk del .build-deps \
    && rm -rf /var/cache/apk/* /tmp/* /usr/share/man /usr/src/php.tar.xz*
```

## More Examples

**Following examples are for non-Alpine images only**. We don't have examples included for the Alpine images.

You can use the image to serve an HTTP/WebSocket server, or run some one-off command with it. e.g.,

```bash
docker run --rm openswoole/swoole "php -m"
docker run --rm openswoole/swoole "php --ri openswoole"
docker run --rm openswoole/swoole "composer --version"
```

We have various examples included under folder "_examples/_" to help developers better use the image. These examples are
numerically ordered. Each example has a _docker-compose.yml_ file included, along with some other files. To run an
example, please start Docker containers using the _docker-compose.yml_ file included, then check HTTP output from URL
http://127.0.0.1 unless otherwise noted. You may use the following commands to start/stop/restart Docker containers:

```bash
./bin/example.sh start   00 # To start container(s) of the first example.
./bin/example.sh stop    00 # To stop container(s) of the first example.
./bin/example.sh restart 00 # To restart container(s) of the first example.
```

To run another example, just replace the last command line parameter _00_ with an example number (e.g., _05_).

Here is a list of the examples under folder "_examples/_":

* Basic examples:
    * **00-autoload**: Restart the Swoole web server automatically if file changes detected under web root.
    * **01-basic**: print out "Hello, World!" using Swoole as backend HTTP server.
    * **02-www**: to use some customized PHP script(s) in the Docker image built.
    * **03-nginx**: to use Swoole behind an Nginx server.
    * **04-entrypoint**: to use a self-defined entrypoint script in the Docker image built.
    * **05-boot**: to update content in the Docker container through a booting script.
    * **06-update-token**: to show how to update server configurations with built-in script _update-token.sh_.
    * **07-disable-default-server**: Please check the [docker-compose.yml](https://github.com/openswoole/docker-swoole/blob/master/examples/07-disable-default-server/docker-compose.yml) file included to see show how to disable the default web server created with _Swoole_.
* Manage PHP extensions and configurations:
    * **10-install-php-extension**: how to install/enable PHP extensions.
    * **11-customize-extension-options**: how to overwrite/customize PHP extension options.
    * **12-php.ini**: how to overwrite/customize PHP options.
    * **13-install-swoole-extension**: Please check the [README](https://github.com/openswoole/docker-swoole/tree/master/examples/13-install-swoole-extension) file included to see how to install Swoole extensions like [postgresql](https://github.com/openswoole/ext-postgresql).
* Manage Supervisord programs:
    * **20-supervisord-services**: to show how to run Supervisord program(s) in Docker.
    * **21-supervisord-tasks**: to show how to run Supervisord program(s) when launching a one-off command with Docker. Please check the [README](https://github.com/openswoole/docker-swoole/tree/master/examples/21-supervisord-tasks) file included to see how to run the example.
    * **22-supervisord-enable-program**: to show how to enable program(s) in Supervisord program.
    * **23-supervisord-disable-program**: to show how to disable Supervisord program(s).
* Debugging:
    * **30-debug-with-gdb**: Please check the [README](https://github.com/openswoole/docker-swoole/tree/master/examples/30-debug-with-gdb) file included to see how to debug your PHP code with _gdb_.
    * **31-debug-with-valgrind**: Please check the [README](https://github.com/openswoole/docker-swoole/tree/master/examples/31-debug-with-valgrind) file included to see how to debug your PHP code with _Valgrind_.
    * **32-debug-with-strace**: Please check the [README](https://github.com/openswoole/docker-swoole/tree/master/examples/32-debug-with-strace) file included to see how to debug your PHP code with _strace_.

# Image Variants

The `openswoole/swoole` images come in three flavors, each designed for a specific use case. **In production environment, we suggest using the Alpine images.**

### 1. `latest`, `<swoole-version>`, and `<swoole-version>-php<php-version>`

* `openswoole/swoole:latest`
* `openswoole/swoole:4.7`
* `openswoole/swoole:4.7-php8.0`
* `openswoole/swoole:4.7.1-php8.0`

This variant is based on the _php:cli_ images, with a few changes. It uses _Supervisord_ to manage booting processes, and has _Composer_ preinstalled.

### 2. `latest-dev`, `<swoole-version>-dev`, and `<swoole-version>-php<php-version>-dev`

* `openswoole/swoole:latest-dev`
* `openswoole/swoole:4.7-dev`
* `openswoole/swoole:4.7-php8.0-dev`
* `openswoole/swoole:4.7.1-php8.0-dev`

This variant is very similar to the previous one, but it has extra tools added for testing, debugging, and monitoring purpose,
including [gdb](https://www.gnu.org/s/gdb), git, lsof, [strace](https://strace.io), [tcpdump](https://www.tcpdump.org), [Valgrind](http://www.valgrind.org), and vim.

### 3. `latest-alpine`, `<swoole-version>-alpine`, and `<swoole-version>-php<php-version>-alpine`

* `openswoole/swoole:latest-alpine`
* `openswoole/swoole:4.7-alpine`
* `openswoole/swoole:4.7-php8.0-alpine`
* `openswoole/swoole:4.7.1-php8.0-alpine`

You can use this variant in the same way as using the _php:alpine_ image, except that we changed the default working directory to _/var/www_.
Also, we have _Composer_ preinstalled in the image.

Note: We don't have development tools built in for Alpine images. There is no Docker images like `openswoole/swoole:4.7.1-php8.0-alpine-dev`.

# Supported Tags and Respective `Dockerfile` Links

## Versioned images (based on stable releases of Swoole)

### Swoole 4.7

| PHP Versions | Default Images | Dev Images | Alpine Images |
|-|-|-|-|
| PHP 8.0 | [4.7.1-php8.0, 4.7-php8.0<br />4.7, latest](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php8.0/cli/Dockerfile) | [4.7.1-php8.0-dev, 4.7-php8.0-dev<br />4.7-dev, latest-dev](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php8.0/cli/Dockerfile) | [4.7.1-php8.0-alpine, 4.7-php8.0-alpine<br />4.7-alpine, latest-alpine](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php8.0/alpine/Dockerfile) |
| PHP 7.4 | [4.7.1-php7.4, 4.7-php7.4](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php7.4/cli/Dockerfile) | [4.7.1-php7.4-dev, 4.7-php7.4-dev](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php7.4/cli/Dockerfile) | [4.7.1-php7.4-alpine, 4.7-php7.4-alpine](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php7.4/alpine/Dockerfile) |
| PHP 7.3 | [4.7.1-php7.3, 4.7-php7.3](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php7.3/cli/Dockerfile) | [4.7.1-php7.3-dev, 4.7-php7.3-dev](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php7.3/cli/Dockerfile) | [4.7.1-php7.3-alpine, 4.7-php7.3-alpine](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php7.3/alpine/Dockerfile) |
| PHP 7.2 | [4.7.1-php7.2, 4.7-php7.2](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php7.2/cli/Dockerfile) | [4.7.1-php7.2-dev, 4.7-php7.2-dev](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php7.2/cli/Dockerfile) | [4.7.1-php7.2-alpine, 4.7-php7.2-alpine](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/4.7.1/php7.2/alpine/Dockerfile) |


## Nightly images (built daily using the master branch of [swoole-src](https://github.com/openswoole/swoole-src))

| PHP Versions | Default Images | Dev Images | Alpine Images |
|-|-|-|-|
| PHP 8.0 | [php8.0](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php8.0/cli/Dockerfile) | [php8.0-dev](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php8.0/cli/Dockerfile) | [php8.0-alpine](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php8.0/alpine/Dockerfile) |
| PHP 7.4 | [php7.4](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php7.4/cli/Dockerfile) | [php7.4-dev](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php7.4/cli/Dockerfile) | [php7.4-alpine](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php7.4/alpine/Dockerfile) |
| PHP 7.3 | [php7.3](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php7.3/cli/Dockerfile) | [php7.3-dev](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php7.3/cli/Dockerfile) | [php7.3-alpine](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php7.3/alpine/Dockerfile) |
| PHP 7.2 | [php7.2](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php7.2/cli/Dockerfile) | [php7.2-dev](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php7.2/cli/Dockerfile) | [php7.2-alpine](https://github.com/openswoole/docker-swoole/blob/master/dockerfiles/latest/php7.2/alpine/Dockerfile) |

# Build Images Manually

The Docker images are built and pushed out automatically through Travis. If you want to build some image manually, please
follow these three steps.

**1**. Install Composer packages. If you have command "composer" installed already, just run `composer update -n`.

**2**. Use commands like following to create dockerfiles:

```bash
./bin/generate-dockerfiles.php latest # Generate dockerfiles to build images from the master branch of Swoole.
./bin/generate-dockerfiles.php 4.7.1  # Generate dockerfiles to build images for Swoole 4.7.1.
```

**3**. Build Docker images with commands like:

```bash
docker build -t openswoole/swoole                     -f dockerfiles/latest/php8.0/cli/Dockerfile   .
docker build -t openswoole/swoole:4.7.1-php8.0        -f dockerfiles/4.7.1/php8.0/cli/Dockerfile    .
docker build -t openswoole/swoole:4.7.1-php8.0-alpine -f dockerfiles/4.7.1/php8.0/alpine/Dockerfile .
```

To build development images (where extra tools are included), add an argument _DEV_MODE_:

```bash
docker build --build-arg DEV_MODE=true -t openswoole/swoole:latest-dev       -f dockerfiles/latest/php8.0/cli/Dockerfile .
docker build --build-arg DEV_MODE=true -t openswoole/swoole:4.7.1-php8.0-dev -f dockerfiles/4.7.1/php8.0/cli/Dockerfile  .
```

# Credits

Current implementation borrows ideas from [Demin](https://github.com/deminy)'s work at [Glu Mobile](https://glu.com).


CHANGELOG.md
Table of Contents
=================

* [Table of Contents](#table-of-contents)
* [Swoole 4.7](#swoole-47)
   * [4.7.1](#471)
      * [Changed](#changed)
   * [4.7.0](#470)
      * [Changed](#changed-1)

# Swoole 4.7

## 4.7.1

### Changed
- Upgrade _Composer_ from 2.1.3 to 2.1.6.


rootfilesystem/entrypoint.sh
#!/usr/bin/env bash

set -e

if [[ ! -z "$@" ]] ; then
    # The container is started to run some one-off command only.
    BOOT_MODE=TASK
else
    # The container is to launch some long running services (e.g., web server, job worker, etc).
    BOOT_MODE=SERVICE
fi
export BOOT_MODE

# Now run .php and .sh scripts under folder /usr/local/boot in order.
boot_scripts=()
shopt -s nullglob
for f in /usr/local/boot/*.sh ; do
    boot_scripts+=("$f")
done
shopt -u nullglob
IFS=$'\n' boot_scripts=($(sort <<<"${boot_scripts[*]}"))
unset IFS
for f in "${boot_scripts[@]}" ; do
    . "$f"
done

# We use option "-c" here to suppress following warning message from console output:
#   UserWarning: Supervisord is running as root and it is searching for its configuration file in default locations...
if [[ "SERVICE" == "${BOOT_MODE}" ]] ; then
    if [[ -n "$(ls /etc/supervisor/conf.d/*.conf 2>/dev/null)" ]] ; then
        /usr/bin/supervisord -c /etc/supervisor/supervisord.conf -n
    else
        tail -f /dev/null
    fi
else
    if [[ -n "$(ls /etc/supervisor/conf.d/*.conf 2>/dev/null)" ]] ; then
        /usr/bin/supervisord -c /etc/supervisor/supervisord.conf
    fi

    if [[ "${1}" =~ ^(ba|)sh$ ]] ; then
        # To support Docker commands like following:
        # docker run --rm openswoole/swoole bash -c "composer --version"
        # docker run --rm openswoole/swoole   sh -c "composer --version"
        exec "$@"
    else
        # To support Docker commands invoked in ECS (via command "aws ecs run-task"), kind of like following:
        # docker run --rm openswoole/swoole "composer --version"
        exec $@
    fi
fi


src/Dockerfile.php
<?php

declare(strict_types=1);

namespace Swoole\Docker;

use Symfony\Component\Yaml\Yaml;
use Twig\Environment;
use Twig\Loader\FilesystemLoader;
use Twig\Error\LoaderError;
use Twig\Error\RuntimeError;
use Twig\Error\SyntaxError;

/**
 * Class Dockerfile
 *
 * @package Swoole\Docker
 */
class Dockerfile
{
    protected const ALPINE = 'alpine';
    protected const CLI = 'cli';

    protected const ALPINE_VERSIONS = [
        // PHP major version => Alpine version,
        '7.2' => '3.12',
        '7.3' => '3.13',
        '7.4' => '3.13',
        '8.0' => '3.13',
        '8.1' => '3.13',
    ];

    protected string $basePath;

    protected string $swooleVersion;

    protected array $config;

    /**
     * Dockerfile constructor.
     *
     * @param string $swooleVersion
     * @throws Exception
     */
    public function __construct(string $swooleVersion)
    {
        $this
            ->setBasePath(dirname(__DIR__))
            ->setSwooleVersion($swooleVersion)
            ->setConfig(Yaml::parseFile("{$this->getConfigFilePath()}"));
    }

    /**
     * @throws Exception
     * @throws LoaderError
     * @throws RuntimeError
     * @throws SyntaxError
     */
    public function render(): void
    {
        foreach ($this->getConfig()['php'] as $phpVersion) {
            foreach ([self::ALPINE, self::CLI] as $type) {
                $this->generateDockerFile($phpVersion, $type, true);
            }
        }
    }

    /**
     * @param string $phpVersion
     * @param string $type
     * @param bool $save
     * @return string
     * @throws Exception
     * @throws LoaderError
     * @throws RuntimeError
     * @throws SyntaxError
     */
    public function generateDockerFile(string $phpVersion, string $type, bool $save = false): string
    {
        $dockerFile = (new Environment(new FilesystemLoader($this->getBasePath())))
            ->load($this->getTemplateFile($type))
            ->render($this->getContext($phpVersion));

        if ($save) {
            $dockerFileDir = $this->getDockerFileDir($type, $phpVersion);
            if (!file_exists($dockerFileDir)) {
                mkdir($dockerFileDir, 0777, true);
            }

            file_put_contents("{$dockerFileDir}/Dockerfile", $dockerFile);
        }

        return $dockerFile;
    }

    /**
     * @return string
     */
    public function getBasePath(): string
    {
        return $this->basePath;
    }

    /**
     * @param string $basePath
     * @return Dockerfile
     * @throws Exception
     */
    public function setBasePath(string $basePath): self
    {
        if (!is_dir($basePath) || !is_readable($basePath)) {
            throw new Exception("base path '{$basePath}' does not point to a directory or not readable");
        }

        $this->basePath = $basePath;

        return $this;
    }

    /**
     * @return string
     */
    public function getSwooleVersion(): string
    {
        return $this->swooleVersion;
    }

    /**
     * @param string $swooleVersion
     * @return Dockerfile
     */
    public function setSwooleVersion(string $swooleVersion): self
    {
        $this->swooleVersion = $swooleVersion;

        return $this;
    }

    /**
     * @return array
     */
    public function getConfig(): array
    {
        return $this->config;
    }

    /**
     * @param array $config
     * @return Dockerfile
     */
    public function setConfig(array $config): self
    {
        $this->config = $config;

        return $this;
    }

    /**
     * @param string $phpVersion
     * @return string
     */
    protected function getPhpMajorVersion(string $phpVersion): string
    {
        return preg_replace('/^(\d+\.\d+).*$/', '$1', $phpVersion);
    }

    /**
     * @param string $type
     * @param string $phpVersion Needed only when creating Dockerfiles for a released version of Swoole.
     * @return string
     */
    protected function getDockerFileDir(string $type, string $phpVersion): string
    {
        return sprintf(
            "%s/dockerfiles/%s/php%s/%s",
            $this->getBasePath(),
            $this->getSwooleVersion(),
            $this->getPhpMajorVersion($phpVersion),
            $type,
        );
    }

    /**
     * @return string
     * @throws Exception
     */
    protected function getConfigFilePath(): string
    {
        $file = $this->getConfigFilePathBySwooleVersion($this->getSwooleVersion());

        if (!is_file($file) || !is_readable($file)) {
            throw new Exception("Unable to load configuration file '{$file}'.");
        }

        return $file;
    }

    /**
     * @param string $swooleVersion
     * @return string
     */
    protected function getConfigFilePathBySwooleVersion(string $swooleVersion): string
    {
        return "{$this->getBasePath()}/config/{$swooleVersion}.yml";
    }

    /**
     * @param string $swooleVersion
     * @return bool
     */
    protected function isValidSwooleVersion(string $swooleVersion): bool
    {
        return (bool) preg_match('/^[1-9]\d*\.(0|[1-9]\d*)\.(0|[1-9]\d*)$/', $swooleVersion);
    }

    /**
     * @see https://github.com/swoole/swoole-src/releases/tag/v4.5.7
     */
    protected function getContext(string $phpVersion): array
    {
        if (
            ($this->getSwooleVersion() === 'latest')
            || (version_compare($this->getSwooleVersion(), '4.6.0-alpha') >= 0)
        ) {
            $optionCurl = ($phpVersion != '8.0.0');
        } else {
            $optionCurl = false;
        }

        if (($this->getSwooleVersion() === 'latest') || (version_compare($this->getSwooleVersion(), '4.5.7') >= 0)) {
            $optionJson = true;
        } else {
            $optionJson = false;
        }

        return array_merge(
            $this->getConfig()['image'],
            [
                'php_version'    => $phpVersion,
                'alpine_version' => $this->getAlpineVersion($phpVersion),
                'swoole_version' => $this->getSwooleVersion(),
                'option_curl'    => $optionCurl,
                'option_json'    => $optionJson,
            ]
        );
    }

    protected function getTemplateFile(string $type): string
    {
        return (self::ALPINE == $type) ? 'Dockerfile.alpine.twig' : 'Dockerfile.cli.twig';
    }

    protected function getAlpineVersion(string $phpVersion): string
    {
        $phpMajorVersion = preg_replace('/^(\d+\.\d+).*$/', '$1', $phpVersion);
        if (!array_key_exists($phpMajorVersion, self::ALPINE_VERSIONS)) {
            throw new Exception("No matching version of Alpine found for PHP {$phpVersion}.");
        }

        return self::ALPINE_VERSIONS[$phpMajorVersion];
    }
}
